### PR TITLE
spidermonkey_91: remove platform restriction

### PR DESCRIPTION
**Pull request by lostnet**
###### Description of changes

Try removing the platform restriction on spidermonkey to build it and couchdb3 on darwin.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by johnjameswhitman** [comment]
Dropping a note that I'm running into this platform restriction when trying to install `libvirt` on darwin right now as well (that package uses [`platforms.unix`](https://github.com/NixOS/nixpkgs/blob/nixos-21.11/pkgs/development/libraries/libvirt/default.nix#L247), but fails to build because of spidermonkey.

**Comment by SuperSandro2000** [comment]
> Dropping a note that I'm running into this platform restriction when trying to install `libvirt` on darwin right now as well (that package uses [`platforms.unix`](https://github.com/NixOS/nixpkgs/blob/nixos-21.11/pkgs/development/libraries/libvirt/default.nix?rgh-link-date=2022-05-31T18%3A37%3A40Z#L247), but fails to build because of spidermonkey.

platforms.unix means it could or did at some point on unix platforms. It does not mean the package could not be broken on darwin throughout time.

Also please follow the contributing guide when naming your commits.

File: pkgs/development/interpreters/spidermonkey/91.nix [hunk 1]
```diff
@@ -120,6 +121,6 @@ stdenv.mkDerivation rec {
     homepage = "https://spidermonkey.dev/";
     license = licenses.mpl20;
     maintainers = with maintainers; [ lostnet ];
-    platforms = platforms.linux;
+    platforms = platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/development/interpreters/spidermonkey/91.nix:124]:
```suggestion
    platforms = platforms.unix;
```

File: pkgs/development/interpreters/spidermonkey/91.nix [hunk 2]
```diff
@@ -39,12 +41,11 @@ stdenv.mkDerivation rec {
     perl
     pkg-config
     python3
-    rust-cbindgen
```
Comment (SuperSandro2000) [inline pkgs/development/interpreters/spidermonkey/91.nix:42]: Is this no longer required?

**Comment by lostnet** [comment]
Hi @SuperSandro2000 this is just an experimental draft since I don't actually have a mac to test on. I was interested in seeing what could be done interactively since that worked remarkably well for the anaconda package manager [pr](https://github.com/conda-forge/staged-recipes/pull/18792), but I wasn't really sure if anyone cared about a dependent of spidermonkey on darwin in nixpkgs, so that's good to know @johnjameswhitman

File: pkgs/development/interpreters/spidermonkey/91.nix (hunk 2 above)
Comment (lostnet) [reply pkgs/development/interpreters/spidermonkey/91.nix:42]: It's not supposed to be necessary according to the [configure](https://github.com/mozilla/gecko-dev/blob/cc03ccd836330833a8d20fa764f780bd5835477e/build/moz.configure/bindgen.configure#L9) and seems to be marked broken in darwin.

**Comment by johnjameswhitman** [comment]
A few quick notes from my end (happy to move this conversation elsewhere if that's preferred):
- I just realized that `libvirt` depends on `spidermonkey_78` so if we figure the issue out here for v91 might be worth replicating for v78.
- Spidermonkey is a transitive dependency for `libvirt` -- is there a good way to find out the path between the two? I tried `nix why-depends nixpkgs.libvirt nixpkgs.spidermonkey-78` but that seems to require building the package, which is blocking it.
- I tried to pull your branch and build `spidermonkey_91` on my local macbook and still got an error, but didn't have time to debug. I'll try and devote some more time to it tomorrow.

```shell
git clone -b spidermac --single-branch git@github.com:lostnet/nixpkgs.git
export NIXPKGS=~/path/to/cloned/nixpkgs
nix-env -f $NIXPKGS -iA spidermonkey_91
```

Which resulted in:
```
...
configurePhase completed in 33 seconds
building
running xcodebuild
error: no project found
builder for '/nix/store/rryspi9aacb12pxwvqxpjf1v44435kjf-spidermonkey-91.10.0.drv' failed with exit code 1
error: build of '/nix/store/rryspi9aacb12pxwvqxpjf1v44435kjf-spidermonkey-91.10.0.drv' failed
```

**Comment by ajs124** [comment]
`nix depends --derivation`. It's through polkit. Which hopefully we'll just transition to `duktape`, once they have a release with support for that out. Although the next release will also have support for `spidermonkey_91` iirc. In general, we're trying to get rid of 78, see #157874.

**Comment by lostnet** [comment]
@johnjameswhitman After switching the xc away from hooks it indeed looked better and I've added libiconv for darwin failing on it for unexplored reasons and it seems to pass. If you can try this out and see if it would work I will look into the specifics of those dependencies and begin addressing @SuperSandro2000 earlier review comments and other things related to getting this out of draft.

**Comment by johnjameswhitman** [comment]
Wooo, it works 🎉 thanks @lostnet!

```
➜  nixpkgs git:(spidermac) nix-shell -I nixpkgs=. -p spidermonkey_91

[nix-shell:~/projects/third_party/github.com/lostnet/nixpkgs]$ ls /nix/store/*spidermonkey*/bin
/nix/store/3wq4apjjlk6dz1hmzza841i72mw18hd3-spidermonkey-91.10.0-dev/bin:
js91-config

/nix/store/ibw7jxj3gwwp56j5hdsfzw3mm6p9jh4r-spidermonkey-91.10.0/bin:
js  js91

[nix-shell:~/projects/third_party/github.com/lostnet/nixpkgs]$ which js
/nix/store/ibw7jxj3gwwp56j5hdsfzw3mm6p9jh4r-spidermonkey-91.10.0/bin/js

[nix-shell:~/projects/third_party/github.com/lostnet/nixpkgs]$ js
js> console.log("hello");
hello
js>
```

**Comment by catap** [comment]
@lostnet do you plan to finish it?

**Comment by lostnet** [comment]
Hi @catap I haven't had much time for spidermonkey recently but will be happy to review https://github.com/NixOS/nixpkgs/pull/179822

**Comment by wegank** [comment]
Closed by #185928.

**Pull request closed without merging**